### PR TITLE
Document disagreements and their resolution

### PR DESCRIPTION
**Pull request by jhoenicke**
Draft for now.

We should figure out the disagreements between cvc5 and Par4, especially where only cvc5 said sat.   Par4 seems to be unsound in both direction (sometimes said sat, sometimes unsat).

### CODE REVIEW

**Comment by jhoenicke** [comment]
some of the benchmarks have a status, maybe we should also mention which

**Comment by tjark** [comment]
> We should figure out the disagreements between cvc5 and Par4, especially where only cvc5 said sat. Par4 seems to be unsound in both direction (sometimes said sat, sometimes unsat).

I would be happy to assist with this. Par4 is a simple wrapper around other SMT solvers, so perhaps there is a soundness bug in one of the wrapped solvers. Is `Single_Query_Track.csv` available?

**Comment by jhoenicke** [comment]
@tjark thanks for your help offer.  The five disagreements of par4 (unsat) with csv5 (sat) are these:

```
UFNIA/2019-Preiner/qf/t3_rw1223.smt2,sat
UFNIA/2019-Preiner/qf/t3_rw1258.smt2,sat
UFNIA/2019-Preiner/qf/t3_rw1283.smt2,sat
UFNIA/2019-Preiner/qf/t3_rw1442.smt2,sat
UFNIA/2019-Preiner/qf/t3_rw805.smt2,sat
```

There are a few other benchmarks where Par4 disagrees with more solvers, but interestingly they are the other way round, so Par4 says (sat) the others say (unsat).  So this seems to be a different bug.

I attached the part of the csv where all disagreements happen (the full Single_Query_Track.csv would be 150 MB, this is only 320kB).

[sq-disagree-excerpt.csv](https://github.com/SMT-COMP/smt-comp/files/6787014/sq-disagree-excerpt.csv)

**Comment by jhoenicke** [comment]
The other Par4 disagreements are these:
```
UFNIA/2019-Preiner/qf/f2_rw235.smt2,unsat
UFNIA/2019-Preiner/qf/f2_rw238.smt2,unsat
UFNIA/2019-Preiner/qf/f2_rw28.smt2,unsat
UFNIA/2019-Preiner/qf/f2_rw29.smt2,unsat
UFNIA/2019-Preiner/qf/f2_rw30.smt2,unsat
UFNIA/2019-Preiner/qf/f2_rw31.smt2,unsat
UFNIA/2019-Preiner/qf/t3_rw1.smt2,unsat
```

**Comment by tjark** [comment]
For UFNIA, Par4 is relying on the 2018 versions of Z3 (version 4.7.1, solver ID 19792 on StarExec) and CVC4 (version master-2018-06-10-b19c840-competition-default, solver ID 19775).

The seven `sat` results for `UFNIA/2019-Preiner/qf/f2_rw235.smt2` etc. are due to Z3. (In my experiments, Z3 timed out on the five remaining benchmarks.)

The five `unsat` results for `UFNIA/2019-Preiner/qf/t3_rw1223.smt2` etc. are due to CVC4, which actually returns `unsat` for **all** of the above benchmarks (including those that Z3 claims to be "sat").

So it is clear that at least one of Z3 and CVC4 (2018 versions) must have a soundness bug for UFNIA.

**Comment by HanielB** [comment]
The model given by `cvc5` for `UFNIA/2019-Preiner/qf/t3_rw805.smt2` is wrong (asserting it back to the formula leads to inconsistency), so that does not help in determining who is wrong.

I did not test for the other `t3_...` but I expect similar results.

**Comment by jhoenicke** [comment]
I looked into the disputed benchmarks and I'm sure they are unsat.

`(intmul k a b)` is defined as `(a * b) % (pow2 k)`  where pow2 is mostly an uninterpreted function, and  `a % 0` is defined as a.
`(intshl k a b)` is defined as `(a * (pow2 b)) % (pow2 k)`. 

For the `%`  (intmodtotal), the rule `(a % x) * (b % x) = (a*b) % x` should hold, so one can pull intmodtotal out.  Then the formulas can be simplified, e.g. for rw1442:
```
(assert (not (forall ((s Int) (t Int) (k Int))
  (=>
   (and (is_bv_var k s) (is_bv_var k t) (is_bv_width k))
   (= (intshl k (intmul k s (intmul k s t)) t) (intmul k s (intmul k s (intshl k t t))))
  )
 )
))
```
both sides of the equality can be simplified to `(s*s*t*(pow2 t)) % (pow2 k)` which shows that the equality is always true.  Hence the implication is true, so the whole asserted formula is false.  Similar reasoning works for all five benchmarks.

**Comment by jhoenicke** [comment]
I'm wondering if we should wait for fixed cvc5 results to see if there are more benchmarks affected.

**Review by aehyvari — APPROVED** [review]
Looks good.

File: 2021/scoring/clean_result_csvs.sh [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+PROCESS_CSV="$SCRIPTDIR/../../tools/process-results/process_results.py"
+RESULT_DIR="$SCRIPTDIR/../results"
+NONINC_DECISION="$SCRIPTDIR/../results/sq-disagreements-decision.csv"
+
+TMPDIR=$(mktemp -d)
+trap "rm -rf $TMPDIR" EXIT
+
+cat blacklist_incremental_*.txt > excluded_incremental.txt
+cat blacklist_nonincremental_*.txt > excluded_nonincremental.txt
```
Comment (aehyvari) [inline 2021/scoring/clean_result_csvs.sh:13]: Where do the `blacklist_incremental_*.txt` and `blacklist_nonincremental_*.txt` come from?

File: 2021/scoring/clean_result_csvs.sh [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+for i in inc mv sq uc; do
+    if [ "$i" == "inc" ]; then
+	EXCLUDED=excluded_incremental.txt
```
Comment (aehyvari) [inline 2021/scoring/clean_result_csvs.sh:17]: These seem to use tabs.

**Comment by aehyvari** [comment]
Both in the parallel and cloud tracks Par4 disagrees with cvc5-gg in the following instances with unknown status in the smt-lib files.  Par4 reports sat and cvc5-gg reports unsat.

```
QF_FP/griggio/fmcad12/test_v5_r10_vr10_c1_s7608.smt2
QF_FP/griggio/fmcad12/test_v7_r12_vr1_c1_s10576.smt2
QF_FP/griggio/fmcad12/test_v7_r7_vr1_c1_s24449.smt2
```

@tjark tried to reproduce the problem in starexec but there Par4 either reports unsat or times out.  Therefore it seems quite clear that these instances are unsatisfiable.  The problem seems to be reproducible in aws infrastructure.  Par4 was similarly affected in a few other benchmarks where starexec status was known.

**Comment by aehyvari** [comment]
> @tjark tried to reproduce the problem in starexec but there Par4 either reports unsat or times out. Therefore it seems quite clear that these instances are unsatisfiable. The problem seems to be reproducible in aws infrastructure. Par4 was similarly affected in a few other benchmarks where starexec status was known.

@tjark updated that this is a file system permission problem where the underlying solver COLIBRI tries to create a temporary file in a directory where the process has no write permission.  The resulting error message is incorrectly interpreted as a SAT response from the solver, causing the apparent disagreements.

**Pull request merged**